chain/client/src/stateless_validation/chunk_validator/mod.rs

```diff
@@ -120,15 +134,66 @@ impl ChunkValidator {
         let chunk_endorsement_tracker = self.chunk_endorsement_tracker.clone();
         let epoch_manager = self.epoch_manager.clone();
         // If we have the chunk extra for the previous block, we can validate the chunk without state witness.
-        // This usually happens when we are a chunk producer
-        let shard_uid = epoch_manager.shard_id_to_uid(chunk_header.shard_id(), &epoch_id)?;
+        // There are two cases here:
+        // 1. We have chunk extra and outgoing receipts cached in `validation_result_cache`. This happens because
+        // we now apply another chunk with the same prev_chunk. This means that at least one chunk produced before
+        // did not get included. Since we have the chunk extra and outgoing receipts, we don't need to validate the
+        // previous state transition again.
+        // 2. We are a chunk producer and therefore have the chunk extra for the previous block saved on disk. We
+        // can also skip validating the chunk state witness in this case.
+        let prev_block = chain.get_block(prev_block_hash)?;
+        let last_header = Chain::get_prev_chunk_header(
+            epoch_manager.as_ref(),
+            &prev_block,
+            chunk_header.shard_id(),
+        )?;
+        let prev_chunk_hash = last_header.chunk_hash();
+        let shard_uid = epoch_manager.shard_id_to_uid(last_header.shard_id(), &epoch_id)?;
+        // Case 1: We have the chunk extra and outgoing receipts cached.
+        let cached_result = {
```

Comment:
this doesn't need to write, right? If it doesn't exist we can just have None?

Reply:
Still need to hold the write lock since it is a lru, but yeah there is no need to insert here.